docs/spec_draft.md

```diff
@@ -102,26 +102,192 @@ never produced as outputs.
 
 ## Programs
 
-StableHLO programs consist of functions. Each function has inputs and outputs
-of supported types and a list of ops in static single-assignment (SSA) form
-which is terminated by a return op which produces the outputs of the function.
-StableHLO ops take inputs and produce outputs.
-
-```mlir
-ml_program.func @example_func(%arg: tensor<2x3xf32>) -> tensor<2x3xf32> {
- %0 = "stablehlo.floor"(%arg) : (tensor<2x3xf32>) -> tensor<2x3xf32>
- %1 = "stablehlo.ceil"(%arg) : (tensor<2x3xf32>) -> tensor<2x3xf32>
- %2 = "stablehlo.add"(%0, %1) : (tensor<2x3xf32>, tensor<2x3xf32>) -> tensor<2x3xf32>
- ml_program.return %2 : tensor<2x3xf32>
+**StableHLO programs** consist of **StableHLO functions**. Each function has
+inputs and outputs of supported types and a list of ops in static
+single-assignment (SSA) form which is terminated by a `return` op which produces
+the outputs of the function.
+
+Here is an example of a program that consists of a function `@main` which takes
+three inputs (`%image`, `%weights` and `%bias`) and produces one output (`%4`).
+Below we describe how this program can be executed.
+
+```mlir
+stablehlo.func @main(
+  %image: tensor<28x28xf32>,
+  %weights: tensor<784x10xf32>,
+  %bias: tensor<1x10xf32>
+) -> tensor<1x10xf32> {
+  %0 = "stablehlo.reshape"(%image) : (tensor<28x28xf32>) -> tensor<1x784xf32>
+  %1 = "stablehlo.dot"(%0, %weights) : (tensor<1x784xf32>, tensor<784x10xf32>) -> tensor<1x10xf32>
+  %2 = "stablehlo.add"(%1, %bias) : (tensor<1x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32>
+  %3 = "stablehlo.constant"() { value = dense<0.0> : tensor<1x10xf32> } : () -> tensor<1x10xf32>
+  %4 = "stablehlo.maximum"(%2, %3) : (tensor<1x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32>
+  "stablehlo.return"(%4): (tensor<1x10xf32>) -> ()
 }
 ```
 
-A program is executed by passing argument values to a given function and
-computing output values. Output values of a function are computed by evaluating
-the graph of ops rooted in the corresponding return op. The evaluation order is
-implementation-defined, as long as ops are evaluated before their uses. Possible
-execution orders of the above example program are `%0` → `%1` → `%2` → `return`
-or `%1` → `%0` → `%2` → `return`.
+## Execution
+
+### Sequential execution
+
+A StableHLO program is executed by providing input values to the `main` function
+and computing output values. Output values of a function are computed by
+executing the graph of ops rooted in the corresponding `return` op.
```

Comment:
Does this mean if the function does not return anything, even if it is invoked nothing would execute?